src/Controls/src/Core/Xaml/TypeConversionExtensions.cs

```diff
@@ -71,42 +72,51 @@ static class TypeConversionExtensions
 		{
 			Func<TypeConverter> getConverter = () =>
 			{
-				MemberInfo memberInfo;
+				if (minfoRetriever != null && minfoRetriever() is MemberInfo memberInfo)
+				{
+					if (memberInfo.CustomAttributes.GetTypeConverterType() is Type converterType)
+					{
+						return (TypeConverter)Activator.CreateInstance(converterType);
+					}
+				}
 
-				var converterTypeName = toType.GetTypeInfo().CustomAttributes.GetTypeConverterTypeName();
-				if (minfoRetriever != null && (memberInfo = minfoRetriever()) != null)
-					converterTypeName = memberInfo.CustomAttributes.GetTypeConverterTypeName() ?? converterTypeName;
-				if (converterTypeName == null && KnownConverterFactories.TryGetValue(toType, out var converterFactory))
+				if (KnownConverterFactories.TryGetValue(toType, out var converterFactory))
 					return converterFactory();
-				if (converterTypeName == null)
-					return null;
-				var convertertype = Type.GetType(converterTypeName);
-				return (TypeConverter)Activator.CreateInstance(convertertype);
+
+				if (!s_converterCache.TryGetValue(toType, out TypeConverter converter))
```

Comment:
The converters we define and provide are reuse-safe and do not have any side effects. We can not guarantee for 3rd party converters.

There are 2 options here:
- only cache our own converters
- require converters to be reusable (I'd go for this)

Reply:
My understanding is that TypeConverters are meant to be cached and reused. They don't have state on them.

[WPF caches them](https://github.com/dotnet/wpf/blob/89d172db0b7a192de720c6cfba5e28a1e7d46123/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Markup/XamlTypeMapper.cs#L3478-L3491).

And so does System.ComponentModel:
1. [It stores them on the ReflectedTypeData class](https://github.com/dotnet/runtime/blob/4ed3063d8bc434fd29fe63cf093c17baa9ff8fd3/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.ReflectedTypeData.cs#L199-L225)
2. [Which is stored in a ReflectTypeDescriptionProvider](https://github.com/dotnet/runtime/blob/4ed3063d8bc434fd29fe63cf093c17baa9ff8fd3/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs#L909-L914)
3. [Which is stored in a static (weak) hashtable](https://github.com/dotnet/runtime/blob/4ed3063d8bc434fd29fe63cf093c17baa9ff8fd3/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs#L1433-L1434)

Reply:
At one point, I was considering using the rest of `System.ComponentModel` to get the converter via `TypeDescriptor.GetConverter`. But I felt this was too heavy-weight to bring in all of the `TypeDescriptor` infrastructure.

Reply:
if 3rd party converters are stageful, it's garbage in, garbage out. works for me